src/dialect.ml

```diff
@@ -0,0 +1,191 @@
+      (Action_dune_lang.run (S.virt __POS__ "ocamlformat")
+         [ S.virt __POS__ (flag_of_kind kind)
+         ; S.virt_var __POS__ "input-file"
+         ])
+  in
+  let file_kind kind extension =
+    { File_kind.
+      kind
+    ; extension
+    ; preprocess = None
+    ; format = Some (Loc.none, format kind, [ ".ocamlformat"
+                                            ; ".ocamlformat-ignore" ])
+    }
+  in
+  let intf = file_kind Ml_kind.Intf ".mli" in
+  let impl = file_kind Ml_kind.Impl ".ml"  in
+  { name       = "ocaml"
+  ; file_kinds = Ml_kind.Dict.make ~intf ~impl
+  }
+
+let reason =
+  let file_kind kind extension =
+    let module S = String_with_vars in
+    let preprocess =
+      Action_dune_lang.run (S.virt __POS__ "refmt")
+        [ S.virt     __POS__ "--print"
+        ; S.virt     __POS__ "binary"
+        ; S.virt_var __POS__ "input-file"
+        ]
+    in
+    let format =
+      Action_dune_lang.run (S.virt __POS__ "refmt")
+        [ S.virt_var __POS__ "input-file"
+        ]
+    in
+    { File_kind.
+      kind
+    ; extension
+    ; preprocess = Some (Loc.none, preprocess)
+    ; format     = Some (Loc.none, format, [])
+    }
+  in
+  let intf = file_kind Ml_kind.Intf ".rei" in
+  let impl = file_kind Ml_kind.Impl ".re"  in
+  { name       = "reason"
+  ; file_kinds = Ml_kind.Dict.make ~intf ~impl
+  }
+
+let ml_suffix { file_kinds = { Ml_kind.Dict.intf ; impl } ; _ } ml_kind =
+  match ml_kind, intf.preprocess, impl.preprocess with
+  | Ml_kind.Intf, None, _ | Impl, _, None -> None
+  | _ -> Some (extension ocaml ml_kind)
+
+module S = struct
+  type dialect = t
+
+  type t =
+    { by_name      : dialect String.Map.t
+    ; by_extension : dialect String.Map.t
+    }
+
+  let empty =
+    { by_name      = String.Map.empty
+    ; by_extension = String.Map.empty
+    }
+
+  let add { by_name ; by_extension } ?loc dialect =
+    let by_name =
+      match String.Map.add by_name dialect.name dialect with
+      | Ok by_name -> by_name
+      | Error _ ->
+        User_error.raise ?loc
+          [ Pp.textf "dialect %S is already defined" dialect.name ]
+    in
+    let add_ext map ext =
+      match String.Map.add map ext dialect with
+      | Ok map ->
+        map
+      | Error dialect ->
+        User_error.raise ?loc
+          [ Pp.textf "extension %S is already registered by dialect %S"
+              ext dialect.name
+          ]
+    in
+    let by_extension =
+      add_ext (add_ext by_extension dialect.file_kinds.intf.extension)
+        dialect.file_kinds.impl.extension
+    in
+    { by_name ; by_extension }
+
+  let of_list dialects =
+    List.fold_left ~f:(add ?loc:None) ~init:empty dialects
+
+  let find_by_name { by_name ; _ } name =
+    String.Map.find by_name name
+
+  let find_by_extension { by_extension ; _ } extension =
+    Option.map ~f:(fun dialect ->
+      let kind =if dialect.file_kinds.intf.extension = extension then
```

Comment:
why don't we check for the impl extension in the else clause?

Reply:
This map is constructed in a way (when calling `Dialect.S.add`) that if an extension is a key then it is the extension of either the interface or the implementation of the corresponding value.